modules/host/network.go

```diff
@@ -49,11 +49,11 @@ func (h *Host) initNetworking(address string) (err error) {
 		return err
 	}
 	// Automatically close the listener when h.tg.Stop() is called.
+	err = h.tg.Add()
+	if err != nil {
+		return err
+	}
 	go func() {
-		err := h.tg.Add()
-		if err != nil {
-			return
-		}
```

Comment:
It's not the `Close` thread that causes the timeout, it's the `threadedHandleConn` thread that timesout. I wrote a test to demonstrate.
```go
func TestHostDeadlockBug(t *testing.T) {
        if testing.Short() {
                t.SkipNow()
        }

        ht, err := blankHostTester("TestHostDeadlockBug")
        if err != nil {
                t.Fatal(err)
        }
        err = ht.host.Close()
        if err != nil {
                t.Fatal(err)
        }

        for i := 0; i < 10000; i++ {
                h, _ := New(ht.cs, ht.tpool, ht.wallet, "localhost:0", filepath.Join(build.SiaTestingDir, "TestHostDeadlockBug", modules.HostDir))
                h.Close()
        }
}
```
And here's the test output
```
$ make test-long pkgs=./modules/host run=TestHostDeadlockBug
rm -rf release doc/whitepaper.aux doc/whitepaper.log doc/whitepaper.pdf
gofmt -s -l -w ./modules/host
go install ./modules/host
go vet ./modules/host
go test -v -race -tags='testing debug' -timeout=300s ./modules/host -run=TestHostDeadlockBug
=== RUN   TestHostDeadlockBug
panic: test timed out after 5m0s

goroutine 2035 [running]:
testing.startAlarm.func1()
        /usr/local/Cellar/go/1.5.3/libexec/src/testing/testing.go:703 +0x16b
created by time.goFunc
        /usr/local/Cellar/go/1.5.3/libexec/src/time/sleep.go:129 +0x6e

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x8a5eb0, 0xb4e820, 0x15, 0x15, 0x1)
        /usr/local/Cellar/go/1.5.3/libexec/src/testing/testing.go:562 +0xafa
testing.(*M).Run(0xc82003ff10, 0xc82014c770)
        /usr/local/Cellar/go/1.5.3/libexec/src/testing/testing.go:494 +0xe5
main.main()
        github.com/NebulousLabs/Sia/modules/host/_test/_testmain.go:94 +0x210

goroutine 17 [syscall, 4 minutes, locked to thread]:
runtime.goexit()
        /usr/local/Cellar/go/1.5.3/libexec/src/runtime/asm_amd64.s:1721 +0x1

goroutine 6 [semacquire, 4 minutes]:
sync.runtime_Semacquire(0xc8200d4ed0)
        /usr/local/Cellar/go/1.5.3/libexec/src/runtime/sema.go:43 +0x26
sync.(*WaitGroup).Wait(0xc8200d4ec4)
        /usr/local/Cellar/go/1.5.3/libexec/src/sync/waitgroup.go:126 +0x118
github.com/NebulousLabs/Sia/sync.(*ThreadGroup).Stop(0xc8200d4ea8, 0x0, 0x0)
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/sync/threadgroup.go:71 +0x10a
github.com/NebulousLabs/Sia/modules/host.(*Host).Close(0xc8200d4a00, 0x0, 0x0)
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/host/host.go:310 +0x7a
github.com/NebulousLabs/Sia/modules/host.TestHostDeadlockBug(0xc82008e1b0)
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/host/host_test.go:119 +0x4dc
testing.tRunner(0xc82008e1b0, 0xb4e8e0)
        /usr/local/Cellar/go/1.5.3/libexec/src/testing/testing.go:456 +0xdd
created by testing.RunTests
        /usr/local/Cellar/go/1.5.3/libexec/src/testing/testing.go:561 +0xaa4

goroutine 7 [sleep]:
time.Sleep(0x4a817c800)
        /usr/local/Cellar/go/1.5.3/libexec/src/runtime/time.go:59 +0xf9
github.com/NebulousLabs/Sia/sync.(*RWMutex).threadedDeadlockFinder(0xc82000bf80)
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/sync/lock.go:78 +0x728
created by github.com/NebulousLabs/Sia/sync.New
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/sync/lock.go:47 +0x127

goroutine 10 [select]:
github.com/NebulousLabs/Sia/modules/gateway.(*Gateway).threadedPeerManager(0xc8200f5030)
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/gateway/peers.go:320 +0x429
created by github.com/NebulousLabs/Sia/modules/gateway.New
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/gateway/gateway.go:169 +0xc36

goroutine 11 [select]:
github.com/NebulousLabs/Sia/modules/gateway.(*Gateway).threadedNodeManager(0xc8200f5030)
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/gateway/nodes.go:135 +0x653
created by github.com/NebulousLabs/Sia/modules/gateway.New
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/gateway/gateway.go:170 +0xc5b

goroutine 12 [IO wait, 4 minutes]:
net.runtime_pollWait(0x2586238, 0x72, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/runtime/netpoll.go:157 +0x63
net.(*pollDesc).Wait(0xc8200f5100, 0x72, 0x0, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/fd_poll_runtime.go:73 +0x56
net.(*pollDesc).WaitRead(0xc8200f5100, 0x0, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/fd_poll_runtime.go:78 +0x44
net.(*netFD).accept(0xc8200f50a0, 0x0, 0x25c0000, 0xc82017e060)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/fd_unix.go:408 +0x2f6
net.(*TCPListener).AcceptTCP(0xc82002c140, 0x0, 0x0, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/tcpsock_posix.go:254 +0x77
net.(*TCPListener).Accept(0xc82002c140, 0x0, 0x0, 0x0, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/tcpsock_posix.go:264 +0x4b
github.com/NebulousLabs/Sia/modules/gateway.(*Gateway).listen(0xc8200f5030)
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/gateway/peers.go:116 +0x5d
created by github.com/NebulousLabs/Sia/modules/gateway.New
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/gateway/gateway.go:173 +0xc80

goroutine 2075 [IO wait, 4 minutes]:
net.runtime_pollWait(0x25857b8, 0x72, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/runtime/netpoll.go:157 +0x63
net.(*pollDesc).Wait(0xc8201de060, 0x72, 0x0, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/fd_poll_runtime.go:73 +0x56
net.(*pollDesc).WaitRead(0xc8201de060, 0x0, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/fd_poll_runtime.go:78 +0x44
net.(*netFD).accept(0xc8201de000, 0x0, 0x25c0000, 0xc8201ce080)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/fd_unix.go:408 +0x2f6
net.(*TCPListener).AcceptTCP(0xc820218000, 0x545d0, 0x0, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/tcpsock_posix.go:254 +0x77
net.(*TCPListener).Accept(0xc820218000, 0x0, 0x0, 0x0, 0x0)
        /usr/local/Cellar/go/1.5.3/libexec/src/net/tcpsock_posix.go:264 +0x4b
github.com/NebulousLabs/Sia/modules/host.(*Host).threadedListen(0xc8200d4a00)
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/host/network.go:178 +0xf7
created by github.com/NebulousLabs/Sia/modules/host.(*Host).initNetworking
        /Users/jordan/golang/src/github.com/NebulousLabs/Sia/modules/host/network.go:94 +0x3f2
exit status 2
FAIL    github.com/NebulousLabs/Sia/modules/host        300.129s
make: *** [test-long] Error 1
```